dev/doc/case-repr.md

```diff
@@ -0,0 +1,115 @@
+the case node can be given as follows.
+
+Provided
+- Γ ⊢ c : Ind@{u} pms Indices
+- Inductive Ind@{i} Δ : forall Θ, Type := cᵢ : forall Ξᵢ, Ind Δ Aᵢ
+- Γ, Θ@{i := u}{Δ := pms} ⊢ p : Type
+- Γ, Ξᵢ@{i := u}{Δ := pms} ⊢ snd brᵢ : p{Θ := Aᵢ{Δ := pms}}
+
+Then Γ ⊢ Case (_, u, pms, ( _, p), _, c, br) : p{Θ := Indices}
+
+In particular, this implies that Γ ⊢ pms : Δ@{i := u}. Parameters are stored in
+the same order as application.
+
+The u universe instance must be a valid instance for the corresponding
+inductive type, in particular their length must coincide.
+
+The `Name.t binder_annot array` appearing both in the return clause and
+in the branches must satisfy some invariants.
+- For branches, it must have the same length as the corresponding Ξᵢ context
+(including let-ins)
+- For the return clause, it must have the same length as the context
+Θ, self : Ind@{u} pms Θ (including let-ins). The last variable appears as
+the term being destructed and correspond to the variable introduced by the
+"as" clause of the user-facing syntax.
+- The relevance annotations must match with the corresponding sort of the
+variable from the context.
+
+Note that the annotated variable array is reversed w.r.t. the context,
+i.e. variables appear in the writing order of the context.
+
+Let-bindings can appear in Δ, Θ or Ξᵢ, since they are arbitrary
+contexts. As a general rule, let bindings appear as binders but not as
+instances. That is, they MUST appear in the variable array, but they MUST NOT
+appear in the parameter array.
+
+Example:
+```
+Inductive foo (X := tt) : forall (Y := X), Type := Foo : forall (Z := X), foo.
+
+Definition case (x : foo) : unit := match x as x₀ in foo with Foo _ z => z end
+```
+The case node of the `case` function is represented as
+```
+Case (
+  _,
+  Instance.empty,
+  [||],
+  ([|(Y, Relevant); (x₀, Relevant)|], unit), (* let (Y := tt) in fun (x₀ : foo) => unit *)
+  NoInvert,
+  #1,
+  [|
+    ([|(z, Relevant)|], #1) (* let z := tt in z *)
+  |]
+)
+```
+
+This choice of representation for let-bindings requires access to the
+environment in some cases, e.g. to compute branch reduction. There is a
+fast-path for non-let-containing inductive types though, which are the vast
+majority.
+
+### Porting plugins
+
+There are two conversion functions from and to the expanded form:
+- `[Inductive, EConstr].expand_case` that goes from the compact to the expanded
+form and cannot fail (assuming the term was well-typed)
+- `[Inductive, EConstr].contract_case` that goes the other way around and will
+raise anomalies if the expanded forms are not fully eta-expanded.
+
+As such, it is alway painless to convert to the old representation. If you go
+the other way around, you must ensure that all the terms you provide the
+compatibility function with are fully eta-expanded, **including let-bindings**.
+Most of the time it happens so, so porting a plugin with 90% correction is
+just a matter of wrapping the plugin code with the two above functions.
```

Comment:
```suggestion
As such, it is always painless to convert to the old representation. Converting
the other way, you must ensure that all the terms you provide the
compatibility function with are fully eta-expanded, **including let-bindings**.
Most of the time it happens so, so porting a plugin with 90% correction is
just a matter of wrapping the plugin code with these two functions.
```

I don't understand "Most of the time it happens so, so porting a plugin with 90% correction".
Do you mean something like "Most of the time, this [the terms are fully eta-expanded] is the case, so 90% of the work is ..."?

Reply:
It's uncommon to generate non-eta expanded branches, except if your plugin does it on purpose. So it means that barring weird corner cases that will sometimes fail *at runtime*, this will work as expected.